chain/events/filter/index.go

```diff
@@ -317,11 +322,111 @@ func (ei *EventIndex) migrateToVersion3(ctx context.Context) error {
+
+	head := chainStore.GetHeaviestTipSet()
+
+	tx, err := ei.db.BeginTx(ctx, nil)
+	if err != nil {
+		return xerrors.Errorf("begin transaction: %w", err)
+	}
+	// Rollback the transaction (a no-op if the transaction was already committed)
+	defer func() { _ = tx.Rollback() }()
+
+	// Alter the event table to add a new column called `emitter`
+	// Note: since `emitter` column does not accept NULL values, set the default to 0 so that the table
+	// can be altered. This means after the migration all the reverted events for which address resolution
+	// cannot be performed will end up with 0 as their emitter actor ID.
+	if _, err = tx.Exec("ALTER TABLE event ADD COLUMN emitter INTEGER NOT NULL DEFAULT 0"); err != nil {
+		return xerrors.Errorf("increment _meta version: %w", err)
+	}
+
+	stmtUpdateEmitterByID, err := tx.Prepare("UPDATE event SET emitter=? WHERE id=?")
+	if err != nil {
+		return xerrors.Errorf("prepare stmtUpdateEmitterByID: %w", err)
+	}
+
+	rows, err := tx.QueryContext(ctx, `SELECT id, emitter_addr FROM event WHERE reverted=false ORDER BY id DESC`)
+	if err != nil {
+		return xerrors.Errorf("select event emitter_addrs: %w", err)
+	}
+
+	for rows.Next() {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		default:
+		}
+
+		var row struct {
+			id          int64
+			emitterAddr []byte
+		}
+		if err := rows.Scan(&row.id, &row.emitterAddr); err != nil {
+			return xerrors.Errorf("read event emitter_addr row: %w", err)
+		}
+		addr, err := address.NewFromBytes(row.emitterAddr)
+		if err != nil {
+			return xerrors.Errorf("parse emitter_addr: %w", err)
+		}
+		emitter, err := actorResolver(ctx, addr, head)
+		if err != nil {
+			return xerrors.Errorf("resolve emitter from addr: %w", err)
+		}
+		updateRes, err := stmtUpdateEmitterByID.ExecContext(ctx, uint64(emitter), row.id)
+		if err != nil {
+			return xerrors.Errorf("resolve emitter from addr: %w", err)
+		}
+		affected, err := updateRes.RowsAffected()
+		if err != nil {
+			return xerrors.Errorf("rows affected: %w", err)
+		}
+		if affected != 1 {
+			log.Warnw("expected exactly one row to be affected as a result of emitter update", "affected", affected)
+		}
+	}
+
+	// Delete event.emitter_addr index introduced in version 3
+	if _, err = tx.ExecContext(ctx, "DROP INDEX IF EXISTS event_emitter_addr"); err != nil {
+		return xerrors.Errorf("drop event_emitter_addr index: %w", err)
+	}
+
+	// Delete the redundant emitter_addr column.
+	if _, err = tx.ExecContext(ctx, "ALTER TABLE event DROP COLUMN emitter_addr"); err != nil {
+		return xerrors.Errorf("drop event.emitter_addr column: %w", err)
+	}
+
+	// Create a new index index fo event.emitter, replacing the deleted event.emitter_addr.
+	if _, err = tx.ExecContext(ctx, createIndexEventEmitter); err != nil {
+		return xerrors.Errorf("create event_emitter index: %w", err)
+	}
+
+	// Increment the schema version in _meta table to 4.
+	if _, err = tx.Exec("INSERT OR IGNORE INTO _meta (version) VALUES (4)"); err != nil {
```

Comment:
do these `OR IGNORE`s make sense to you for the versions in the migrations? I think I'd want to know if I got here and it was already at `4`, even if it was an error log and not fatal.

Reply:
It keeps to the existing convention that's present across previous schema migrations. Apart from that removing `OR IGNORE` here will not cause the migration to fail. Right?

Other than that, the scenario you described is not currently possible through the code executed by Lotus. Unless the user modifies the DB externally while the migration is running or runs two instances of lotus at the same time.